h2/src/test/org/h2/test/synth/TestConcurrentUpdate.java

```diff
@@ -29,107 +27,91 @@
-                            case 0:
-                                insert.setInt(1, x);
-                                insert.setString(2, data);
-                                insert.execute();
-                                break;
-                            case 1:
-                                update.setString(1, data);
-                                update.setInt(2, x);
-                                update.execute();
-                                break;
-                            case 2:
-                                delete.setInt(1, x);
-                                delete.execute();
-                                break;
-                            case 4:
-                                select.setInt(1, x);
-                                ResultSet rs = select.executeQuery();
-                                while (rs.next()) {
-                                    rs.getString(2);
+            Task[] tasks = new Task[THREADS];
+            for (int i = 0; i < THREADS; i++) {
+                final int threadId = i;
+                Task t = new Task() {
+                    @Override
+                    public void call() throws Exception {
+                        Random r = new Random(threadId);
+                        try (Connection conn = getConnection(url)) {
+                            PreparedStatement insert = conn.prepareStatement(
+                                    "merge into test values(?, ?)");
+                            PreparedStatement update = conn.prepareStatement(
+                                    "update test set name = ? where id = ?");
+                            PreparedStatement delete = conn.prepareStatement(
+                                    "delete from test where id = ?");
+                            PreparedStatement select = conn.prepareStatement(
+                                    "select * from test where id = ?");
+                            while (!stop) {
+                                int x = r.nextInt(ROW_COUNT);
+                                String data = "x" + r.nextInt(ROW_COUNT);
+                                switch (r.nextInt(3)) {
+                                    case 0:
+                                        insert.setInt(1, x);
+                                        insert.setString(2, data);
+                                        insert.execute();
+                                        break;
+                                    case 1:
+                                        update.setString(1, data);
+                                        update.setInt(2, x);
+                                        update.execute();
+                                        break;
+                                    case 2:
+                                        delete.setInt(1, x);
+                                        delete.execute();
+                                        break;
+                                    case 4:
```

Comment:
`case 4:` is not executed at all. And was not executed before these changes, because `nextInt(3)` may return only 0, 1, or 2. If this test is changed for some reason may be this should be fixed too?

Reply:
Absolutely will do. Our brain and vision works in a mysterious way - it only now occurred to be that cases are not sequentially numbered either. Time to go to elementary school along with Thomas. :smile: